### DISCUSSION TITLE
How to install necessary and custom tools on the EC2 runner?

### DISCUSSION THREAD

**Discussion by machulav** [machulav/ec2-github-runner#18] [Ideas]
In one of the commit, @snussik added a comment, that could be very important for many users of the action. 
So I've decided to create a new discussion for that and try to figure out what to do with that.

> @machulav glad, that you continue maintaining this action!
> In my opinion it's a bad idea to use some definite AMI. For example docker installation for Amazon Linux 2 according to their off.docs is following:
> sudo yum update -y
> sudo amazon-linux-extras install docker
> sudo service docker start
> sudo usermod -a -G docker ec2-user
> 
> As for my case I also need docker-compose -> it should be installed separately. And so on.
> 
> My idea is that its much more simpler to create ami and config it properly as I need with one command:
> sudo amazon-linux-extras install -y docker && sudo service docker start && sudo usermod -a -G docker ec2-user && sudo chkconfig docker on && sudo yum install -y git && sudo curl -L https://github.com/docker/compose/releases/download/1.22.0/docker-compose-$(uname -s)-$(uname -m) -o /usr/local/bin/docker-compose && sudo chmod +x /usr/local/bin/docker-compose && docker-compose version
> 
> and then just use it with your action.

The original comment you can find [here](https://github.com/machulav/ec2-github-runner/commit/816e24caf6144c981430dd70dc17b6b56e25666b#commitcomment-45746189).

**Comment by machulav** [comment]
@snussik can you please answer the questions below to let me better understand your point?

> In my opinion it's a bad idea to use some definite AMI.
1. What exactly do you mean by "definite AMI"?
2. Do you think so only because you want to install your own custom tools on the AMI and you think preparing the AMI beforehand would be the best option? Or because you think that it would be helpful to support also some other Linux distributions like Ubuntu and so on?

**Reply by snussik** [reply]
@machulav the main idea of my previous comment is that it's not the best workflow to mix several operations in one product (github action). You action makes a cool thing: gives a possibility to create a remote self-hosted runner and execute in as one of the steps.
Speaking about instances, they always different and depend on the task. For example: i need docker with preconfigured cpu and memory limits, or additionally I want to have htop and ctop for monitoring.
For all those and other tasks it's better to use special insrtuments: terraform, ansible or manual creation of image with needed specs. And even simpler: different actions tasks from [Action Marketplace](https://github.com/marketplace?query=node)

When you try to inject come commands in instance initialisation process - it always causes some bugs on different productions.
So, the answers to your questions are:
1. No need to base on definite AMIs like Amazon Linux 2 and so on.  In your case now it should be Linux x64 based, according to [github docs](https://docs.github.com/en/free-pro-team@latest/actions/hosting-your-own-runners/about-self-hosted-runners#supported-operating-systems-for-self-hosted-runners) and your code. But when you try to install something with yum, apt and so on - you narrows implementation of the action.
2. Thats why I think that it's a good idea to give users opportunity to create AMIs themselves from scratch. It'll prevent your action from various bug issues connected with operating systems.

**Reply by machulav** [reply]
@snussik The main idea behind installing all the necessary tools in the initialization phase is simplifying the user setup process. So the user doesn't even need to think about the instance pre-configuration - the action will automatically install and configure everything.

However, as I see now, this approach has more cons than pros:
- it's not flexible enough, as you've already described;
- EC2 instance creation takes 10-30 seconds longer as the action has to run the init script every time the instance is created.

So, I agree with you on this point. Those two issues above seem to have more sense in the long perspective than a small convenience of the automatic configuration.

I think I will update this in the new version of the action and remove the init script from it, which installs all the necessary tools.

**Reply by machulav** [reply]
Create an issue for that: #20

**Reply by machulav** [reply]
@snussik FYI: the issue is released in the latest version.

Thank you fur the suggestion!
